### PR TITLE
labelmaker: external API calls

### PR DESCRIPTION
**Pull request by bnewbold**
Making calls out to APIs to help with labeling determinations.

This round of work will include a basic (not production ready) NSFW image filter, and SQRL API. Everything is still single-threaded, not concurrent.

### CODE REVIEW

**Comment by bnewbold** [comment]
Rebased this and it is ready for review.

I tested with https://github.com/bluesky-social/indigo/pull/67 cherry-picked on top of this and it worked; this branch passes tests but won't actually work for blobs until that fix lands.

The makefile changes in https://github.com/bluesky-social/indigo/pull/65 are also helpful for testing this code.

File: labeling/micro_nsfw_img.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+		return nil, fmt.Errorf("micro-NSFW-img request failed  statusCode=%d", res.StatusCode)
+	}
+
+	respBytes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read micro-NSFW-img resp body: %v", err)
+	}
+
+	var nsfwScore MicroNSFWImgResp
+	if err = json.Unmarshal(respBytes, &nsfwScore); err != nil {
```
Comment (whyrusleeping) [inline labeling/micro_nsfw_img.go:103]: best practice to do `if err := json....` here, that way you arent accidentally changing an external scope value

File: labeling/service.go [hunk 2]
```diff
@@ -129,33 +147,181 @@ func (s *Server) wantAnyRecords(ctx context.Context, ra *events.RepoAppend) bool
...
 	log.Infof("labeling record: %v", uri)
 	var labelVals []string
+	var blobs []lexutil.Blob
 	switch nsid {
 	case "app.bsky.feed.post":
 		post, suc := rec.(*appbsky.FeedPost)
 		if !suc {
 			return []string{}, fmt.Errorf("record failed to deserialize from CBOR: %s", rec)
 		}
+		/* XXX(bnewbold): debugging broken post record
```
Comment (whyrusleeping) [inline labeling/service.go:173]: delete?

File: labeling/service.go [hunk 3]
```diff
@@ -129,33 +147,181 @@ func (s *Server) wantAnyRecords(ctx context.Context, ra *events.RepoAppend) bool
...
+			}
+		}
 	case "app.bsky.actor.profile":
 		profile, suc := rec.(*appbsky.ActorProfile)
 		if !suc {
 			return []string{}, fmt.Errorf("record failed to deserialize from CBOR: %s", rec)
 		}
+		// XXX(bnewbold): debugging broken profile record
+		profileJson, _ := json.Marshal(profile)
+		log.Infof("labeling profile: %v", string(profileJson))
```
Comment (whyrusleeping) [inline labeling/service.go:209]: cleanup if it works now

File: labeling/service.go [hunk 4]
```diff
@@ -129,33 +147,181 @@ func (s *Server) wantAnyRecords(ctx context.Context, ra *events.RepoAppend) bool
...
+		}
+		if profile.Banner != nil {
+			blobs = append(blobs, *profile.Banner)
+		}
+	}
+
+	log.Infof("will process %d blobs", len(blobs))
+	for _, blob := range blobs {
+		if blob.Cid == "" {
+			return []string{}, fmt.Errorf("received stub blob (CID undefined)")
```
Comment (whyrusleeping) [inline labeling/service.go:241]: any particular reason not to just return nil? In an error case you arent going to use the value

File: labeling/service.go [hunk 5]
```diff
@@ -129,33 +147,181 @@ func (s *Server) wantAnyRecords(ctx context.Context, ra *events.RepoAppend) bool
...
+		labelVals = append(labelVals, blobLabels...)
+	}
+	return dedupeStrings(labelVals), nil
+}
+
+func (s *Server) downloadRepoBlob(ctx context.Context, did string, blob *lexutil.Blob) ([]byte, error) {
+	var blobBytes []byte
+
+	if blob.Cid == "" {
+		return []byte{}, fmt.Errorf("invalid blob to download (CID undefined)")
```
Comment (whyrusleeping) [inline labeling/service.go:269]: similarly, just return `nil, err`

File: labeling/service.go [hunk 6]
```diff
@@ -129,33 +147,181 @@ func (s *Server) wantAnyRecords(ctx context.Context, ra *events.RepoAppend) bool
...
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return []byte{}, fmt.Errorf("failed to fetch blob from PDS. did=%s cid=%s statusCode=%d", did, blob.Cid, resp.StatusCode)
+	}
+
+	blobBytes, err = io.ReadAll(resp.Body)
+	if err != nil {
+		return blobBytes, nil
```
Comment (whyrusleeping) [inline labeling/service.go:293]: return the error

File: labeling/util.go [hunk 7]
```diff
@@ -45,3 +45,16 @@ func loadKey(kfile string) (*did.PrivKey, error) {
 		Type: out,
 	}, nil
 }
+
+// NOTE(bnewbold): I don't know of a more idiomatic way to do this in golang
+func dedupeStrings(in []string) []string {
```
Comment (whyrusleeping) [inline labeling/util.go:50]: this is fairly typical. If its performance sensitive or in a fast path there are better ways, but simple is better

File: labeling/service.go (hunk 2 above)
Comment (bnewbold) [reply labeling/service.go:173]: whoops!

File: labeling/service.go (hunk 6 above)
Comment (bnewbold) [reply labeling/service.go:293]: good catch!

File: labeling/util.go (hunk 7 above)
Comment (bnewbold) [reply labeling/util.go:49]: removed comment

File: labeling/micro_nsfw_img.go (hunk 1 above)
Comment (bnewbold) [reply labeling/micro_nsfw_img.go]: updated this an similar cases

File: labeling/service.go (hunk 4 above)
Comment (bnewbold) [reply labeling/service.go]: updated this and similar cases

**Comment by bnewbold** [comment]
Updated all comments. Good eye on the fat-fingered incorrect returns in error handling!

File: labeling/keyword_labeler.go [hunk 8]
```diff
@@ -1,42 +1,60 @@
...
+
+	jsonFile, err := os.Open(fpath)
+	if err != nil {
+		return kwl, fmt.Errorf("failed to load JSON file: %v", err)
+	}
+	defer jsonFile.Close()
+
+	raw, err := io.ReadAll(jsonFile)
+	if err != nil {
+		return kwl, fmt.Errorf("failed to load JSON file: %v", err)
```
Comment (whyrusleeping) [inline labeling/keyword_labeler.go:52]: definitely not blocking, but its always a bit awkward when i see a non-nil value along with an error return. It signals to me that you intend to use the error *and* the value, when the intention here is to ignore the value in the error case. It draws the eye unnecessarily

**Review by whyrusleeping — APPROVED** [review]

**Pull request merged**